lucet-concurrency-tests/tests/killswitch.rs

```diff
@@ -0,0 +1,731 @@
+use lucet_runtime::vmctx::Vmctx;
+use lucet_runtime::{
+    lucet_hostcall, Error, Instance, InstanceHandle, KillError, KillSuccess, Limits, Region,
+    RunResult, TerminationDetails, TrapCode,
+};
+use std::sync::Arc;
+use std::thread;
+
+use lucet_module::FunctionPointer;
+use lucet_runtime::MmapRegion;
+use lucet_runtime_internals::lock_testpoints::{SyncWaiter, Syncpoint};
+use lucet_runtime_internals::module::Module;
+use lucet_runtime_internals::module::{MockExportBuilder, MockModuleBuilder};
+use lucet_runtime_internals::vmctx::lucet_vmctx;
+use lucet_runtime_tests::build::test_module_c;
+use lucet_runtime_tests::helpers::test_ex;
+use lucet_runtime_tests::helpers::test_nonex;
+
+static mut ENTERING_GUEST: Option<Syncpoint> = None;
+
+/// A convenience wrapper around running our mock timeout module's `onetwothree` function,
+/// and asserting that it returned the expected result.
+fn run_onetwothree(inst: &mut Instance) {
+    let retval = inst
+        .run("onetwothree", &[])
+        .expect("instance runs")
+        .unwrap_returned();
+    assert_eq!(libc::c_int::from(retval), 123);
+}
+
+pub fn mock_killswitch_module() -> Arc<dyn Module> {
+    extern "C" fn onetwothree(_vmctx: *mut lucet_vmctx) -> std::os::raw::c_int {
+        123
+    }
+
+    extern "C" fn run_guest(_vmctx: *mut lucet_vmctx) {
+        unsafe {
+            // ENTERING_GUEST is only populated if the test requires this syncpoint be checked.
+            if let Some(entering_guest) = ENTERING_GUEST.as_ref() {
+                entering_guest.check();
+            }
+        }
+    }
+
+    extern "C" fn infinite_loop(_vmctx: *mut lucet_vmctx) {
+        unsafe {
+            // ENTERING_GUEST is only populated if the test requires this syncpoint be checked.
+            if let Some(entering_guest) = ENTERING_GUEST.as_ref() {
+                entering_guest.check();
+            }
+        }
+        loop {}
+    }
+
+    extern "C" fn fatal(vmctx: *mut lucet_vmctx) {
+        extern "C" {
+            fn lucet_vmctx_get_heap(vmctx: *mut lucet_vmctx) -> *mut u8;
+        }
+
+        unsafe {
+            let heap_base = lucet_vmctx_get_heap(vmctx);
+
+            // Using the default limits, each instance as of this writing takes up 0x200026000 bytes
+            // worth of virtual address space. We want to access a point beyond all the instances,
+            // so that memory is unmapped. We assume no more than 16 instances are mapped
+            // concurrently. This may change as the library, test configuration, linker, phase of
+            // moon, etc change, but for now it works.
```

Comment:
:new_moon: :crescent_moon: :full_moon:

wonderful comments, as always.

Reply:
this is an @acfoltzer special! Much of this is a straight copy of `lucet-runtime-tests`' `guest_fault`, with the timeout mock module mixed in. I'm actually on the fence about keeping several of these hostcalls because there's a fair bit not used in these tests.

Reply:
This is actually adapted from a comment in the original C runtime's tests, so probably a @pchickey special originally! :cowboy_hat_face: